src/pages/data-first-and-data-last-a-comparison.md

```diff
@@ -0,0 +1,474 @@
+
+The advantages of the data-first approach when it comes to editor integration is something that language designers with a vast experience like Anders Hejlsberg —creator and lead architect of TypeScript— [have pointed out in the past](https://github.com/Microsoft/TypeScript/issues/15680#issuecomment-307571917).
+
+### Intuitive design for functions with multiple params
+
+One of the downsides of the data-last approach is that sometimes it makes harder to understand what a function with two operands is doing.
+
+For example, the `String.concat` function in OCaml standard library:
+
+```reason
+let foo = String.concat("a", ["b", "c"])
+```
+
+Because we are used to left-to-right reading —like the inference engine— we could guess the resulting value of `foo` is the string `"abc"`, but it's actually `"bca"` because the API is data-last.
+
+Another example is the `Js.String.split` in BuckleScript:
+
+```reason
+let bar = Js.String.split("a|b|c", "|")
+```
+
+The resulting value is an array with one element, `["|"]` instead of `["a", "b", "c"]`. The [`Js.String` module](https://bucklescript.github.io/bucklescript/api/Js.String.html) in BuckleScript was originally designed to follow the data-last approach as well.
+
+So, if we are not using currying and the pipe operator, we have to read the parameters "backwards" when dealing with data-last functions, which is not very intuitive.
+
+### Performance
+
+Pipe first operator `->` is implemented as purely syntactic sugar, as mentioned above. This means that, from the compiler perspective, the usage of `->` means that no extra functions calls are involved.
+
+This is not what happens with the pipe last operator `|>`, that gets compiled into a function call. While the compiler does a lot of optimizations to inline values whenever possible, this difference makes it hard to optimize in some specific cases.
+
+Here is an example of a binding to a JavaScript function using BuckleScript interop capabilities that shows how this can impact the resulting code.
+
+With pipe last, using `[@bs.send.pipe]` annotations:
+
+```reason
+[@bs.send.pipe: user]
+external update: (~isAdmin: bool=?, ~age: int) => user = "";
+
+let jane = jane |> update(~age=45);
+```
+
+Which results in the following generated JavaScript code:
+
+```javascript
+var jane$1 = (function(param) {
+  return function(param$1) {
+    return param$1.update(
+      param !== undefined ? Caml_option.valFromOption(param) : undefined,
+      45
+    );
+  };
+})(undefined)(jane);
+```
+
+Now, with pipe first:
+
+```reason
+[@bs.send]
+external update: (user, ~isAdmin: bool=?, ~age: int, unit) => user = "";
+
+let jane = jane->update(~age=45, ());
+```
+
+Results in this JavaScript code:
+
+```javascript
+var jane$1 = jane.update(undefined, 45);
+```
+
+Quite simpler! We went from two function definitions and a curried application in both of them, to a one-liner with just one function call.
```

Comment:
It would be good to add here that the complexity of the data-last version's output is really coming from the optional parameter. In other words, BuckleScript is not being able to uncurry the function call specifically because it needs to check for the presence of `isAdmin`. If you don't have this optional param, the outputs are identically simple.